### DISCUSSION TITLE
Random discussion and questions about RetroPhone

### DISCUSSION THREAD

**Discussion by ktownsend-personal** [ktownsend-personal/RetroPhone#2] [General]
```
╭────────────────────────────────────────────────────────────────────────────╮
│ __________        __               __________.__                           │
│ \______   \ _____/  |________  ____\______   \  |__   ____   ____   ____   │
│  |       _// __ \   __\_  __ \/  _ \|     ___/  |  \ /  _ \ /    \_/ __ \  │
│  |    |   \  ___/|  |  |  | \(  <_> )    |   |   Y  (  <_> )   |  \  ___/  │
│  |____|_  /\___  >__|  |__|   \____/|____|   |___|  /\____/|___|  /\___  > │
│         \/     \/                                 \/            \/     \/  │
│                            2023, Keith Townsend                            │
│              https://github.com/ktownsend-personal/RetroPhone              │
╰────────────────────────────────────────────────────────────────────────────╯
```

## Welcome!
I encourage questions, ideas and general discussion about this project.

**Comment by ondfrukt** [comment]
Hi,

I am into a similar project and also have this dream to build an intercom system with old pulse-dialing phones and/or DTMF phones. It all started with two Swedish military phones (Ericsson M/37) that could establish a link between each other and send ring pulses with a hand crank through the line, but it would be so fun to have more than just two subscribers in my small intercom. I started to look for an exchange of some sort but gave up after a year or two. However, the idea to build my own exchange was growing slowly and actually, with no experience in programming or electronics, I managed to create a mockup based on this concept: [https://www.youtube.com/watch?v=J0ZVI14qiLo&list=LL&index=1&ab_channel=theswedenmovie]. I learned a lot but the solutions had too many bugs and were not reliable enough for my now rising demands, and I am now into using the SLIC module as you and @GadgetReboot do. I have, unlike you two, not the knowledge and therefor not the ambition to create a trunk line to another intercom network and will stay with just a local, which is more than enough for me and my ability.

But now to my question; I'm kind of stuck in how I will connect the several phones within the same exchange. Maybe you have some thoughts about how to solve that? I just can't figure out how to connect all _A_out_ from several SLIC modules with the demanded _A_in_ and vice versa? And is it possible to ring up a third phone during a call or just enable this third phone to pick up its handset and join the conversation?

Finaly I must thank you for sharing your work and thought, it has inspired me a lot and creates a great desire to learn more in programming and electronics in general. Thanks!

**Reply by GadgetReboot** [reply]
I'm working (slowly) on an analog signal routing matrix that will have several phone jacks to allow plugging in several stand-alone SLIM boards (a SLIM contains a SLIC module and other hardware to represent a stand alone central office, and each phone plugs into its own SLIM board).
The SLIM has an extra phone jack to interconnect with another SLIM to route audio directly via the tip and ring between boards, but in this case it will connect to the new routing matrix phone jack.  The way the SLIM is designed, that extra jack is only enabled when the phone is off hook so it only carries audio level signals and should be safe to route through the switch matrix, and we avoid having to make extra connections for the SLIC audio in/out paths.

Then there will be an ESP module on the routing matrix, and it will become a wireless trunk "master" node, so when one of the phone SLIM boards wants to make a call, it is sending the request to the ESP on the new matrix board, which keeps track of which nodes are available or busy.  If a call can be routed, it will connect those two SLIM boards via the matrix and it will behave the same as the prototype where two SLIMs were directly wired with those extra phone jacks.

The advantage will be multiple simultaneous calls can be routed between separate nodes, or extra nodes can be connected into an existing signal path for a multi node party line, if it all works out.

I'm hoping to have that sometime in the coming months but it takes a lot of attention to get the software going, which isn't my strong point but it'll happen.

**Reply by ktownsend-personal** [reply]
The key to an exchange is having a matrix circuit and a way to instruct it which nodes to connect. This is something I've talked to @GadgetReboot about quite a lot. I haven't gotten to the point of actually building that part, but there are a couple of ways to go about it. 

The two basic options revolve around a grid of switches, either physical relay switches or a matrix microchip that does essentially the same thing as a grid of relays. You would need as many switches as you have phones multiplied by the number of concurrent separate connections. For example, if I have 4 phones and want to support two separate conversations that would be 8 switches. In this diagram I have two conversations connected (1 & 3, 2 & 4):

||phone 1|phone 2|phone 3|phone 4|
|---|---|---|---|---|
|channel 1|x||x||
|channel 2||x||x|

In this example, we have a 3-way call:

||phone 1|phone 2|phone 3|phone 4|
|---|---|---|---|---|
|channel 1|x|x|x||
|channel 2|||||

Of course, the real work is controlling the switches. Are you using a controller for each SLIC, or one controller for all SLIC modules?  In my project I have been assuming one ESP32 per SLIC, and probably a separate ESP32 to manage the switching matrix. If putting all of the SLIC modules into one box, it would be feasible to control them all with one controller, in which case the SLIC and matrix management can be done in one software routine. I hadn't considered that approach, but I like the simplicity of having just one controller. I should give that more thought. In my current design, separate ESP32's requires SPI or something similar between all of the controllers to coordinate channel selection when making a call. 

Even though I've contemplated this matrix idea a lot, I'm actually hoping to get the audio to work over WiFi and then I can avoid the matrix and connect point to point over WiFi. That will make 3-way calling challenging, but I may not implement it.

If you're curious about the matrix microchip, @GadgetReboot has some videos where he experiments with one to connect guitar effects in different configurations. It's pretty cool.
